aten/src/ATen/core/op_registration/kernel_functor.h

```diff
@@ -0,0 +1,178 @@
+      (void)std::initializer_list<int>{(
+        torch::jit::push(*stack, return_type_to_ivalue(std::move(std::get<indices>(output))))
+      , 0)...};
+    }
+  };
+
+  template<class KernelFunctor, class Enable = void> struct wrap_kernel_functor final {};
+
+  // SFINAE version for kernels that return an output
+  template<class KernelFunctor>
+  struct wrap_kernel_functor<KernelFunctor, guts::enable_if_t<!std::is_same<void, typename guts::infer_function_traits_t<KernelFunctor>::return_type>::value>> final {
+    static void call(Stack* stack, KernelCache* cache) {
+      static_assert(std::is_base_of<OperatorKernel, KernelFunctor>::value, "Kernel functor must inherit from c10::OperatorKernel");
+
+      constexpr size_t num_inputs = guts::infer_function_traits_t<KernelFunctor>::number_of_parameters;
+      KernelFunctor* functor = static_cast<KernelFunctor*>(cache);
+      auto output = call_functor_with_ivalue_args<KernelFunctor>(functor, torch::jit::last(*stack, num_inputs));
+      torch::jit::pop(*stack, num_inputs);
+      push_outputs<typename guts::infer_function_traits_t<KernelFunctor>::return_type>::call(std::move(output), stack);
+    }
+  };
+
+  // SFINAE version for kernels that don't return an output
+  template<class KernelFunctor>
+  struct wrap_kernel_functor<KernelFunctor, guts::enable_if_t<std::is_same<void, typename guts::infer_function_traits_t<KernelFunctor>::return_type>::value>> final {
+    static void call(Stack* stack, KernelCache* cache) {
+      static_assert(std::is_base_of<OperatorKernel, KernelFunctor>::value, "Kernel functor must inherit from c10::OperatorKernel");
+
+      constexpr size_t num_inputs = guts::infer_function_traits_t<KernelFunctor>::number_of_parameters;
+      KernelFunctor* functor = static_cast<KernelFunctor*>(cache);
+      call_functor_with_ivalue_args<KernelFunctor>(functor, torch::jit::last(*stack, num_inputs));
+      torch::jit::pop(*stack, num_inputs);
+    }
+  };
+
+  template<class KernelFunctor, class... Args>
+  class KernelFactory final {
+  public:
+    explicit constexpr KernelFactory(Args... args)
+    : constructor_parameters_(std::move(args)...) {}
+
+    std::unique_ptr<KernelCache> operator()() const {
+      return guts::apply(
+        [] (const Args&... params) {return guts::make_unique<KernelFunctor>(params...); },
+        constructor_parameters_);
+    }
+
+  private:
+    std::tuple<Args...> constructor_parameters_;
+  };
+}
+
+/**
+ * Use this to register an operator whose kernel is implemented as a functor
+ *
+ * Example:
+ *
+ * > namespace {
+ * >   class my_kernel_cpu final : public c10::OperatorKernel {
+ * >   public:
+ * >     Tensor operator()(Tensor a, Tensor b) {...}
+ * >   };
+ * > }
+ * >
+ * > static auto registry = c10::RegisterOperators()
+ * >     .op("my_op",
+ * >         c10::kernel<my_kernel_cpu>(),
+ * >         c10::dispatchKey(CPUTensorId()));
+ *
+ * The functor constructor can take arguments to configure the kernel.
```

Comment:
it's an interesting feature. What's the use case for it? Those arguments would be the same per-schema, so it'd be equivalent to making those params as template arguments to the op kernel implementation (but with smaller code size)

Reply:
It has a smaller code size and it also allows a wider range of values. Template arguments can only be primitive types, no std::string or Tensors for example. And template arguments must be defined at compile time, while this can take runtime values.

The only current use case for it is that the deprecated runtime function pointer API can easily be implemented with it - by having a kernel that takes the runtime function pointer as a constructor argument and stores it in a member. That is, however, an implementation detail of that API and could be done without having constructor parameters of kernels be a public feature.

But I imagine there will be other uses for this, even though I don't have a concrete one in mind right now.